sycl/include/CL/sycl/detail/context_impl.hpp

```diff
@@ -58,7 +58,7 @@ class context_impl {
   RT::PiContext &getHandleRef();
   const RT::PiContext &getHandleRef() const;
 
-  std::map<OSModuleHandle, RT::PiProgram> &getCachedPrograms() {
+  std::map<KernelSetId, RT::PiProgram> &getCachedPrograms() {
```

Comment:
They are not. Is case (2) supposed to be supported? If so, grouping kernels into sets using `OSModuleHandle` as well as kernel names would resolve this issue.

Reply:
Sorry if I'm late.
Why don't  we want to search for KernelName in "other" os modules?

Reply:
>  Is case (2) supposed to be supported?

Yes.

> How would we choose which OSModule to pick given a kernel name that's present in both?

The algorithm is first get OSModuleHandle instance for the kernel name's string address (which is always part of static data section of the corresponding module), then use only device code modules belonging to this OSModuleHandle instance to resolve the kernel. The OSModuleHandle <-> device code modules correspondence is established during OS module loading when it registers its device modules within the SYCL runtime.

> Why don't we want to search for KernelName in "other" os modules?

That would be true dynamic device code linkage, which is not supported today. When it is supported, static linkage will still be needed anyway, and it would be the common case actually.

For true dynamic device code linkage a lot of design is needed - symbol visibility rules, device module compile-time identification mechanism, compile-time dependence specification between modules, new resolution mechanism for cross-OSModule visible symbols.

Reply:
I didn't get part about linkage.
Current state is that we don't support linking in runtime, so we can work with self-contained modules only. But how does it contradict with looking for program registered from another module if all of them self contained?

Reply:
> But how does it contradict with looking for program registered from another module if all of them self contained?

My logic is: call into another OSModule is a dynamic linkage, which is not designed and not supported for the device code. So, until it is designed/supported we have only static linkage which means we can call into this OSModule's kernels.

Today the only way for users to have cross-OSModule device code linkage is to manually manage device code modules and use cl::sycl::program APIs for linkage and building.

Reply:
I don't consider it as a linkage(in fact there is no link operation on device code involved). But removing OSModule would enable the feature and simplify the code - we have one key instead of two.  KernelName -> image instead of ModuleOS->KernelName->image

Reply:
Simplicity is not the only factor here.
- What happens if two OS modules contain kernel with the same name?
- What is the kernel module search algorithm - go through all the OS modules and all the device modules within each OS module?

Reply:
> What happens if two OS modules contain kernel with the same name?

If I read the SYCL specification correctly it says the KernelName should be unique identifier in the program. So, there should be no problems for a proper program.
Also we could add diagnostic if the names are not unique.

> What is the kernel module search algorithm - go through all the OS modules and all the device modules within each OS module?

We have two global maps that contain information about all kernels/images for the program. So, we just need to find corresponding "kernel set id" for a KernelName given and then using it find a correct image. P.S. This is how Sergey is doing in the patch.

Reply:
> If I read the SYCL specification correctly it says the KernelName should be unique identifier in the program

I don't think SYCL spec describes these scenarios in necessary details. A practical case, however, if an application uses a 3rd-party module which can contain arbitrarily named kernel, and name conflict is possible.

Reply:
In theory yes, but this concern also valid for regular functions from regular C++ world. And usually people solve this problem by putting such functions into namespace.

Reply:
@kbobrovs Added support for this case by grouping kernels based on both OSModule and kernel names to avoid conflicts.